packages/protocol/contracts/team/airdrop/MerkleClaimable.sol

```diff
@@ -11,25 +11,41 @@ import { OwnableUpgradeable } from
 import { MerkleProofUpgradeable } from
     "lib/openzeppelin-contracts-upgradeable/contracts/utils/cryptography/MerkleProofUpgradeable.sol";
 
-import { Proxied } from "../../common/Proxied.sol";
-
 /// @title MerkleClaimable
 /// Contract for managing Taiko token airdrop for eligible users
-// TODO(dani): add claimStart and claimEnd timestamp so claim() can only be
-// called between these two timestamps.
-// Better to add non-reentrance guard to claim().
-
 abstract contract MerkleClaimable is OwnableUpgradeable {
     mapping(bytes32 => bool) public isClaimed;
     bytes32 public merkleRoot;
+    uint128 public claimStart;
+    uint128 public claimEnd;
 
     event Claimed(bytes32 hash);
 
+    error CLAIM_NOT_ONGOING();
     error CLAIMED_ALREADY();
     error INVALID_PROOF();
+    error INVALID_MERKLE_ROOT();
+
+    modifier ongoingClaim() {
+        if (
+            merkleRoot == 0x0 || claimStart == 0 || claimEnd == 0
+                || claimStart > block.timestamp || claimEnd < block.timestamp
+        ) revert CLAIM_NOT_ONGOING();
+        _;
+    }
+
+    function claim(
+        bytes calldata data,
+        bytes32[] calldata proof
+    )
+        external
+        ongoingClaim
+    {
+        (address user, uint256 amount) = abi.decode(data, (address, uint256));
```

Comment:
can we not assume the data is going to follow specical format/encoding? It should be up to the sub-contract to decide.